### PR TITLE
[Feature] CarStreaming benchmark that inspects values.

### PR DESCRIPTION
**Pull request by ruseinov**
## Summary of changes

Changes introduced in this pull request:

- Introduces a benchmark that inspects every Ipld data of every block, to set reasonable upper throughput limit expectations.

## Reference issue to close (if applicable)

Work on #3314 

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by ruseinov** [comment]
Streaming `forest_snapshot_calibnet_2023-08-02_height_788380.forest.car`

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| stream inspect | 7.955 ± 0.297 | 7.686 | 8.414 | 10.00 ± 0.58 |
| stream | 0.796 ± 0.035 | 0.749 | 0.845 | 1.00 |

File: src/tool/subcommands/benchmark_cmd.rs [hunk 1]
```diff
@@ -110,6 +122,34 @@ async fn benchmark_car_streaming(input: Vec<PathBuf>) -> Result<()> {
...
+    let mut s = Box::pin(
+        futures::stream::iter(input)
+            .then(File::open)
+            .map_ok(BufReader::new)
+            .and_then(CarStream::new)
+            .try_flatten(),
+    );
+    while let Some(block) = s.try_next().await? {
+        let block: Block = block;
+        let ipld: Ipld = from_slice_with_fallback(&block.data).unwrap_or(Ipld::Null);
```
Comment (lemmih) [inline src/tool/subcommands/benchmark_cmd.rs:146]: You can check if the data is DAG_CBOR before parsing it. Then, if it fails to parse, you can propagate the error.

`if cid.codec() == DAG_CBOR { ... }`

**Comment by ruseinov** [comment]
Streaming `forest_snapshot_calibnet_2023-08-02_height_788380.forest.car.zst`
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| stream inspect | 11.616 ± 0.642 | 11.259 | 12.761 | 3.02 ± 0.17 |
| stream | 3.845 ± 0.023 | 3.816 | 3.872 | 1.00 |

**Comment by ruseinov** [comment]
Inspect Streaming vs DFS `forest_snapshot_calibnet_2023-08-02_height_788380.forest.car.zst`
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| inspect stream | 11.381 ± 0.142 | 11.250 | 11.608 | 1.00 |
| dfs | 26.395 ± 0.640 | 26.057 | 27.536 | 2.32 ± 0.06 |

**Comment by ruseinov** [comment]
Inspect Streaming vs DFS `forest_snapshot_calibnet_2023-08-02_height_788380.forest.car`
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| stream inspect | 7.828 ± 0.062 | 7.762 | 7.904 | 1.00 |
| dfs | 11.206 ± 0.255 | 11.059 | 11.660 | 1.43 ± 0.03 |

File: src/utils/db/car_stream.rs [hunk 2]
```diff
@@ -130,7 +130,7 @@ impl<ReaderT: AsyncBufRead> Stream for CarStream<ReaderT> {
                 bytes.advance(data_offset as usize);
                 Ok(Block {
                     cid,
-                    data: bytes.to_vec(),
+                    data: bytes.into(),
```
Comment (lemmih) [inline src/utils/db/car_stream.rs:133]: These two lines both copy the bytes.

**Review by lemmih — CHANGES_REQUESTED** [review]
Benchmark looks fine. Can merge once the change in car_stream.rs is either reverted or explained.

File: src/tool/subcommands/benchmark_cmd.rs [hunk 3]
```diff
@@ -111,6 +124,36 @@ async fn benchmark_car_streaming(input: Vec<PathBuf>) -> Result<()> {
...
+async fn benchmark_car_streaming_inspect(input: Vec<PathBuf>) -> Result<()> {
+    let mut sink = indicatif_sink("traversed");
+    let ipld_to_cid = |ipld| {
+        if let Ipld::Link(cid) = ipld {
+            return Some((cid, ()));
+        }
+        None
+    };
+
+    let mut s = Box::pin(
```
Comment (aatifsyed) [inline src/tool/subcommands/benchmark_cmd.rs:139]: why not pin to the stack?

**Review by aatifsyed — APPROVED** [review]

File: src/tool/subcommands/benchmark_cmd.rs (hunk 3 above)
Comment (ruseinov) [reply src/tool/subcommands/benchmark_cmd.rs:139]: Just a copy-paste from the original bench without inspection. Good point though, I'm planning to perhaps unify the two benches going forward.

**Pull request merged**